File: dp/1048.java
class Solution {
    public int longestStrChain(String[] words) {
        // 按string length构造dp, 查询每个word的sub word length, 当前word的length
        // 为sub word length+1。遍历所有word即可找到最大长度
        int ret = 0;
        Arrays.sort(words, new Comparator<String>(){
            public int compare(String s1, String s2) {
                return s1.length() - s2.length();
            }
        });
        Map<String, Integer> dp = new HashMap<>();
        for (String word : words) {
            int cur = 0;
            for (int i = 0; i < word.length(); i++) {
                String pre = word.substring(0, i) + word.substring(i+1);
                cur = Math.max(cur, dp.getOrDefault(pre, 0)+1);
            }
            dp.put(word, cur);
            ret = Math.max(ret, cur);
        }
        return ret;
    }
}

File: hashtable/1048.cc
class Solution {
public:
    int longestStrChain(vector<string>& words) {
        std::sort(words.begin(), words.end(), [](std::string &s1, std::string &s2){
            return s1.size() < s2.size();
        });
        std::unordered_map<std::string, int> m;
        int ret = 0;
        for (const auto &word : words) {
            std::string temp = "";
            int cur = 0;
            for (int i = 0; i < word.size(); ++i) {
                temp = word.substr(0, i) + word.substr(i+1);
                auto iter = m.find(temp);
                if (iter != m.end()) {
                    cur = std::max(cur, iter->second + 1);
                } else {
                    cur = std::max(cur, 1);
                }
            }
            m[word] = cur;
            ret = std::max(ret, cur);
        }
        return ret;
    }
};
